File: app/src/main/java/com/example/lpiem/magiccards/MainActivity.java
package com.example.lpiem.magiccards;

import android.content.Intent;

import android.net.Uri;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.Toast;

import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.widget.Button;

import com.facebook.AccessToken;
import com.facebook.CallbackManager;
import com.facebook.FacebookCallback;
import com.facebook.FacebookException;
import com.facebook.GraphRequest;
import com.facebook.GraphResponse;
import com.facebook.login.LoginManager;
import com.facebook.login.LoginResult;
import com.facebook.login.widget.LoginButton;

import org.json.JSONObject;

import java.util.Arrays;


import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInApi;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.common.api.ApiException;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;

public class MainActivity extends AppCompatActivity implements View.OnClickListener{


    private GoogleSignInClient mGoogleSignInClient;

    private static final int RC_SIGN_IN = 9001;


    private LoginButton loginButton;
    private CallbackManager callbackManager;
    private AccessToken accessToken;

    @Override
    protected void onCreate(Bundle savedInstanceState) {

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);



        // Configure sign-in to request the user's ID, email address, and basic
// profile. ID and basic profile are included in DEFAULT_SIGN_IN.
        GoogleSignInOptions gso = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                .requestEmail()
                .build();

        // Build a GoogleSignInClient with the options specified by gso.
        mGoogleSignInClient = GoogleSignIn.getClient(this, gso);

        // Check for existing Google Sign In account, if the user is already signed in
// the GoogleSignInAccount will be non-null.
        GoogleSignInAccount account = GoogleSignIn.getLastSignedInAccount(this);
        Toast.makeText(this, "account",Toast.LENGTH_SHORT).show();

        findViewById(R.id.sign_in_button).setOnClickListener(this);
        findViewById(R.id.button_sign_out).setOnClickListener(this);
        findViewById(R.id.button_disconnect).setOnClickListener(this);



        callbackManager = CallbackManager.Factory.create();


        accessToken = AccessToken.getCurrentAccessToken();
        boolean isLoggedIn = accessToken != null && !accessToken.isExpired();

        if (isLoggedIn){

            getUserDetails(accessToken);

        }

        loginButton = (LoginButton) findViewById(R.id.login_button);
        loginButton.setReadPermissions("email");

        // Callback registration
        loginButton.registerCallback(callbackManager, new FacebookCallback<LoginResult>() {
            @Override
            public void onSuccess(LoginResult loginResult) {


                getUserDetails(loginResult.getAccessToken());

                Log.e("succesLogButton","yes");
            }

            @Override
            public void onCancel() {
                Log.e("succesLogButtonC","no");
            }

            @Override
            public void onError(FacebookException exception) {
                Log.e("succesLogButtonE","no");
            }
        });



    }


    @Override
    protected void onResume() {
        super.onResume();

        GoogleSignInAccount acct = GoogleSignIn.getLastSignedInAccount(this);
        if (acct != null) {
            String personName = acct.getDisplayName();
            String personGivenName = acct.getGivenName();
            String personFamilyName = acct.getFamilyName();
            String personEmail = acct.getEmail();
            Log.d("PersonalInfo", "person email: "+personEmail);
            String personId = acct.getId();
            Log.d("PersonalInfo", "person Name: "+personName);
            Uri personPhoto = acct.getPhotoUrl();
        }
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.sign_in_button:
                signIn();
                break;

            // ...
            case R.id.button_sign_out:
                signOut();
                break;
// ...

            case R.id.button_disconnect:
                revokeAccess();
                break;
        }

    }

    private void signIn() {
        Intent signInIntent = mGoogleSignInClient.getSignInIntent();
        startActivityForResult(signInIntent, RC_SIGN_IN);
    }

    private void signOut() {
        Log.d("connect", "signOut: ");
        mGoogleSignInClient.signOut()
                .addOnCompleteListener(this, new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {
                        // ...
                    }
                });
    }

    private void revokeAccess() {
        Log.d("connect", "revokeAccess: ");
        mGoogleSignInClient.revokeAccess()
                .addOnCompleteListener(this, new OnCompleteListener<Void>() {
                    @Override
                    public void onComplete(@NonNull Task<Void> task) {
                        // ...
                    }
                });
    }




    private void handleSignInResult(Task<GoogleSignInAccount> completedTask) {
        try {
            GoogleSignInAccount account = completedTask.getResult(ApiException.class);

            // Signed in successfully, show authenticated UI.
            Toast.makeText(this,"Connecte",Toast.LENGTH_SHORT).show();
            this.goToMenuActivity();
        } catch (ApiException e) {
            // The ApiException status code indicates the detailed failure reason.
            // Please refer to the GoogleSignInStatusCodes class reference for more information.
            Log.w("Connect", "signInResult:failed code=" + e.getStatusCode());
            Toast.makeText(this,"Error",Toast.LENGTH_SHORT).show();
        }
    }

    private void goToMenuActivity (){
        Intent intent = new Intent(MainActivity.this, MenuActivity.class);
        GoogleSignInAccount acct = GoogleSignIn.getLastSignedInAccount(this);
        if (acct != null) {
            intent.putExtra("ACCOUNT", acct);
            startActivity(intent);
        }

    }



    protected void getUserDetails(AccessToken loginResult) {
        GraphRequest data_request = GraphRequest.newMeRequest(
                loginResult, new GraphRequest.GraphJSONObjectCallback() {
                    @Override
                    public void onCompleted(
                            JSONObject json_object,
                            GraphResponse response) {
                        Intent intent = new Intent(MainActivity.this, MenuActivity.class);
                        intent.putExtra("userProfile", json_object.toString());
                        startActivity(intent);
                    }

                });
        Bundle permission_param = new Bundle();
        permission_param.putString("fields", "id,name,email,picture.width(120).height(120)");
        data_request.setParameters(permission_param);
        data_request.executeAsync();

    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        callbackManager.onActivityResult(requestCode, resultCode, data);
        super.onActivityResult(requestCode, resultCode, data);

        if (requestCode == RC_SIGN_IN) {
            // The Task returned from this call is always completed, no need to attach
            // a listener.


            Task<GoogleSignInAccount> task = GoogleSignIn.getSignedInAccountFromIntent(data);
            handleSignInResult(task);
        }
    }

}

File: app/src/main/java/com/example/lpiem/magiccards/MainActivity.kt
package com.example.lpiem.magiccards

import Managers.UserManager
import Models.User
import android.content.Intent
import android.os.Bundle
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import com.facebook.AccessToken
import com.facebook.FacebookCallback
import com.facebook.FacebookException
import com.facebook.login.LoginManager
import com.facebook.login.LoginResult
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInAccount
import com.google.android.gms.common.api.ApiException
import com.google.android.gms.tasks.Task
import controllers.InterfaceCallBackController
import controllers.MagicCardRetrofitController
import kotlinx.android.synthetic.main.activity_main.*
import viewModel.ConnexionViewModel


class MainActivity : AppCompatActivity() , InterfaceCallBackController {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        ConnexionViewModel.initialize(this,this)


        if(ConnexionViewModel.isLoggedIn!!) {
            ConnexionViewModel.getUserDetails(AccessToken.getCurrentAccessToken())
        }

        login_button.setReadPermissions("email")

        //Bouton login de facebook
        login_button.registerCallback(ConnexionViewModel.callbackManager, object : FacebookCallback<LoginResult> {
            override fun onSuccess(loginResult: LoginResult) {

                ConnexionViewModel.getUserDetails(loginResult.accessToken)

                Log.e("succesLogButton", "yes")
            }

            override fun onCancel() {
                Log.e("succesLogButtonC", "no")
            }

            override fun onError(exception: FacebookException) {
                Log.e("succesLogButtonE", "no")
            }
        })

        //butons pour le compte google
        sign_in_button.setOnClickListener { ConnexionViewModel.signIn(this) }
    }



    override fun onResume() {
        super.onResume()

        val acct = GoogleSignIn.getLastSignedInAccount(this)
        if (acct != null) {
            val personName = acct.displayName
            val personGivenName = acct.givenName
            val personFamilyName = acct.familyName
            val personEmail = acct.email

 //           Log.d("PersonalInfo", "person email: " + personEmail!!)
            val personId = acct.id
 //           Log.d("PersonalInfo", "person Name: " + personName!!)
            val personPhoto = acct.photoUrl
        }
    }

    private fun signIn() {
        val signInIntent = ConnexionViewModel.mGoogleSignInClient!!.getSignInIntent()
        startActivityForResult(signInIntent, ConnexionViewModel.RC_SIGN_IN)
    }

    private fun signOut() {
        Log.d("connect", "signOut: ")
        ConnexionViewModel.mGoogleSignInClient!!.signOut()
                .addOnCompleteListener(this) {
                    // ...
                }
    }

    private fun revokeAccess() {
        Log.d("connect", "revokeAccess: ")
        ConnexionViewModel.mGoogleSignInClient!!.revokeAccess()
                .addOnCompleteListener(this) {
                    // ...
                }
    }


    private fun handleSignInResult(completedTask: Task<GoogleSignInAccount>) {
        try {
            val account = completedTask.getResult(ApiException::class.java)

            // Signed in successfully, show authenticated UI.
            Toast.makeText(this, "Connecte", Toast.LENGTH_SHORT).show()
            this.goToMenuActivity()
        } catch (e: ApiException) {
            // The ApiException status code indicates the detailed failure reason.
            // Please refer to the GoogleSignInStatusCodes class reference for more information.
            Log.w("Connect", "signInResult:failed code=" + e.statusCode)
            Toast.makeText(this, "Error", Toast.LENGTH_SHORT).show()
        }

    }

    private fun goToMenuActivity() {

        ConnexionViewModel.acctGoogle = GoogleSignIn.getLastSignedInAccount(this)
        if (ConnexionViewModel.acctGoogle != null) {
            connexionToTheAppWithGoogle(ConnexionViewModel.acctGoogle!!.id.toString())
        }

    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        ConnexionViewModel.signInResult(requestCode, resultCode, data, this)
    }


    override fun onWorkDone(result: Any) {
        if (result is Map<*, *>) {
            if (result["google"] === true) {
                try {

                    val intent = Intent(this@MainActivity, SplashScreenActivity::class.java)
                    startActivity(intent)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            } else if (result["facebook"] === true) {
                try {
                    val intent = Intent(this@MainActivity, SplashScreenActivity::class.java)
                    startActivity(intent)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            } else if (result["google"] === false){
                inscriptionGoogleAccount(UserManager.user!!)
                ConnexionViewModel.insription!!.observe(this, Observer {
                    Log.d("NewUserId", UserManager.user!!.googleId!!)
                    connexionToTheAppWithGoogle(UserManager.user!!.googleId!!)
                })

            } else if (result["facebook"] === false) {
                inscriptionFacebookAccount(UserManager.user!!)
                ConnexionViewModel.insription!!.observe(this, Observer {
                    ConnexionViewModel.connexionToTheAppWithFacebook(UserManager.user!!.fbId!!)
                })
            }

        }
    }


    private fun connexionToTheAppWithGoogle(googleId: String) {
        val controller = MagicCardRetrofitController(this )
        controller.callUserGoogleId(googleId, UserManager.user!!)
    }

    private fun inscriptionGoogleAccount(user: User) {
        val controller = MagicCardRetrofitController(this )
        controller.createGoogleUser(User(null,ConnexionViewModel.acctGoogle!!.id.toString(),null,ConnexionViewModel.acctGoogle!!.displayName.toString()));
    }

    private fun inscriptionFacebookAccount(user: User) {
        val controller = MagicCardRetrofitController(this )
        controller.createFacebookUser(User(null, null,ConnexionViewModel.acctFacebook!!.get("id").toString(), ConnexionViewModel.acctFacebook!!.get("name").toString()));
    }

}
